evm/src/cpu/kernel/stack_manipulation.rs

```diff
@@ -0,0 +1,259 @@
+#[derive(Eq, PartialEq)]
+struct Node {
+    stack: Vec<StackItem>,
+    cost: u32,
+}
+
+impl PartialOrd for Node {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Node {
+    fn cmp(&self, other: &Self) -> Ordering {
+        // We want a min-heap rather than the default max-heap, so this is the opposite of the
+        // natural ordering of costs.
+        other.cost.cmp(&self.cost)
+    }
+}
+
+/// Like `StackReplacement`, but without constants or macro vars, since those were expanded already.
+#[derive(Eq, PartialEq, Hash, Clone, Debug)]
+enum StackItem {
+    NamedItem(String),
+    Literal(Literal),
+}
+
+#[derive(Clone, Debug)]
+enum StackOp {
+    Push(Literal),
+    Pop,
+    Dup(u8),
+    Swap(u8),
+}
+
+/// A set of candidate operations to consider for the next step in the path from `src` to `dst`.
+fn next_ops(src: &[StackItem], dst: &[StackItem], unique_literals: &[Literal]) -> Vec<StackOp> {
+    if let Some(top) = src.last() && !dst.contains(top) {
+        // If the top of src doesn't appear in dst, don't bother with anything other than a POP.
+        return vec![StackOp::Pop]
+    }
+
+    let mut ops = vec![StackOp::Pop];
+
+    ops.extend(
+        unique_literals
+            .iter()
+            // Only consider pushing this literal if we need more occurrences of it, otherwise swaps
+            // will be a better way to rearrange the existing occurrences as needed.
+            .filter(|lit| {
+                let item = StackItem::Literal((*lit).clone());
+                let src_count = src.iter().filter(|x| **x == item).count();
+                let dst_count = dst.iter().filter(|x| **x == item).count();
+                src_count < dst_count
+            })
+            .cloned()
+            .map(StackOp::Push),
+    );
+
+    let src_len = src.len() as u8;
+
+    ops.extend(
+        (1..=src_len)
+            // Only consider duplicating this item if we need more occurrences of it, otherwise swaps
+            // will be a better way to rearrange the existing occurrences as needed.
+            .filter(|i| {
+                let item = &src[src.len() - *i as usize];
+                let src_count = src.iter().filter(|x| *x == item).count();
+                let dst_count = dst.iter().filter(|x| *x == item).count();
+                src_count < dst_count
+            })
+            .map(StackOp::Dup),
+    );
+
+    ops.extend((1..src_len).map(StackOp::Swap));
+
+    ops
+}
+
+impl StackOp {
+    fn cost(&self) -> u32 {
+        let (cpu_rows, memory_rows) = match self {
+            StackOp::Push(n) => {
+                let bytes = n.to_trimmed_be_bytes().len() as u32;
+                // This is just a rough estimate; we can update it after implementing PUSH.
+                (bytes, bytes)
+            }
+            Pop => (1, 1),
+            StackOp::Dup(_) => (1, 2),
+            StackOp::Swap(_) => (1, 4),
```

Comment:
nit: Maybe add a small explanation for these numbers.

Reply:
Done 👍